### PR TITLE
Add an actions to deploy binaries automatically

### LINKED ISSUE #12: Publish static Linux binaries with releases
Hi, thanks for the awesome tool!

I'd like to use this in a CI build. The build would go faster and use fewer resources if I were able to download a statically-linked Linux executable from the Github releases, rather than use `cargo install`. It's actually pretty easy to set up Github Actions to automatically create draft releases with attached static binaries whenever you push a git tag. An example CI setup from one of my own projects is available here: https://github.com/langston-barrett/mdlynx/blob/main/.github/workflows/release.yml

Feel free to close this issue if this doesn't appeal to you or sounds like too much work, I just figured I'd share my use-case and the above link in case it's helpful!

### PR DESCRIPTION
**Pull request by Luni-4**
This PR adds an action to deploy binaries automatically and it fixes #12

### CODE REVIEW

**Comment by RalfJung** [comment]
Thanks for the PR!

That's a lot of stuff I don't know much about, I'll have to figure out how to best review this. Are these binaries created by `cargo build` really portable enough to be suited for binary releases?

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        with:
+          path: ./binaries
+
+      - name: Create a release
+        uses: softprops/action-gh-release@v1
+        with:
+          name: v${{ steps.tagName.outputs.tag }}
+          files: |
+            ./binaries/**/*.zip
+            ./binaries/**/*.tar.gz
```
Comment (RalfJung) [inline .github/workflows/deploy.yml:144]: Why are these wrapped in a tarball/zipfile? They are just a single file, and e.g. for #12 it seems easier of that file is immediately the binary.

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      - uses: actions/checkout@v3
+
+      - name: Install Rust stable
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          toolchain: stable
+
+      - name: Create Cargo.lock
+        run: |
+          cargo update
```
Comment (RalfJung) [inline .github/workflows/deploy.yml:125]: We have a lockfile in the repo, why is this needed?

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+      - name: Get version
+        id: tagName
+        run: |
+          VERSION=$(cargo pkgid | cut -d# -f2 | cut -d: -f2)
+          echo "tag=$VERSION" >> $GITHUB_OUTPUT
```
Comment (RalfJung) [inline .github/workflows/deploy.yml:131]: This job only runs when we push a tag, right? Shouldn't it use that tag's name?

Also I don't understand the `:` part of the `cut` chain here. Doesn't `cargo pkgid | cut -d# -f2` already give the right answer?

**Comment by Luni-4** [comment]
> Thanks for the PR!
> 
> That's a lot of stuff I don't know much about, I'll have to figure out how to best review this. Are these binaries created by `cargo build` really portable enough to be suited for binary releases?

Thanks for your review! 

I've mainly created these binaries for CI, to reduce `cargo-careful`'s installing time, I don't know if they are suited for binary releases, I suppose they could be used for that purpose too, but not sure about that

**Comment by RalfJung** [comment]
> I've mainly created these binaries for CI,

So after we land this, could you test whether they actually work on all 3 targets?

**Comment by Luni-4** [comment]
> > I've mainly created these binaries for CI,
> 
> So after we land this, could you test whether they actually work on all 3 targets?

Sure! I'm going to report that here, but let's see if it works the deployment task first :)

**Comment by Luni-4** [comment]
@RalfJung Can we land this PR or do I need to make some other changes?

**Comment by RalfJung** [comment]
Currently it is waiting for me to have time for review. I do this in my spare time and there are other PRs before this one in my queue. I will get to it eventually :)

File: .github/workflows/deploy.yml [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+      - name: Download artifacts
+        uses: actions/download-artifact@v2
+        with:
+          path: ./binaries
+
+      - name: Create a release
+        uses: softprops/action-gh-release@v1
+        with:
+          name: v${{ steps.tagName.outputs.tag }}
```
Comment (RalfJung) [inline .github/workflows/deploy.yml:130]: It looks like this defaults to the name of the tag... so we could probably avoid those `tagName` jobs by just removing this entirely? The tag name should always match the cargo package version for a release.

If anything we might want to put a little check somewhere which ensures that the tag name and the cargo version match.

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+
+    - name: Install musl
+      if: contains(matrix.target, 'linux-musl')
+      run: |
+        sudo apt-get install musl-tools
+
+    - name: Build cargo-careful
+      run: |
+        # TODO: Remember to add RUSTFLAGS=+crt-static for musl target when
+        # static linkage will not be the default behaviour
```
Comment (RalfJung) [inline .github/workflows/deploy.yml:76]: How would I even do that? Sounds quite tedious to do that with GHA.

File: .github/workflows/deploy.yml (hunk 4 above)
Comment (Luni-4) [reply .github/workflows/deploy.yml:130]: This parameter assigns a name to a release which corresponds to `v{tag}`, Do you want this part completely removed or use the cargo package version?

File: .github/workflows/deploy.yml (hunk 5 above)
Comment (Luni-4) [reply .github/workflows/deploy.yml:76]: Since this action has been automatically generated from a tool, we can remove this comment because it's just a reminder

File: .github/workflows/deploy.yml (hunk 4 above)
Comment (RalfJung) [reply .github/workflows/deploy.yml:128]: What I am saying is that, reading the docs for `softprops/action-gh-release`, the `name` field defaults to the current tag. So what is the point of assigning it explicitly here?

File: .github/workflows/deploy.yml (hunk 5 above)
Comment (RalfJung) [reply .github/workflows/deploy.yml]: Are you sure of this? It sounds like in the future this will default to linking the CRT dynamically which would be an issue for distributable binaries.

File: .github/workflows/deploy.yml (hunk 4 above)
Comment (Luni-4) [reply .github/workflows/deploy.yml]: Oh, you're perfectly right.
- Removed that field
- Add a check to compare crate tag and `Github` tag

File: .github/workflows/deploy.yml (hunk 5 above)
Comment (Luni-4) [reply .github/workflows/deploy.yml]: Yep, better to be explicit, so link CRT statically

**Comment by RalfJung** [comment]
Thanks! I have tweaked the code a bit, should be good to go now. And then I guess I should test whether it works. :)

File: .github/workflows/release.yml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        target: ${{ matrix.target }}
+
+    - name: Install musl
+      if: contains(matrix.target, 'linux-musl')
+      run: |
+        sudo apt-get install musl-tools
+
+    - name: Build cargo-careful
+      run: |
+        RUSTFLAGS=+crt-static cargo build --release --target ${{ matrix.target }}
```
Comment (RalfJung) [inline .github/workflows/release.yml:69]: This is wrong, looks like it has not been tested. :/ So the 0.3.1 is butched... I hope 0.3.2 will work...

The correct flag is `RUSTFLAGS="-C target-feature=+crt-static" cargo build ...`

File: .github/workflows/release.yml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+    - name: Strip binary
+      run: |
+        strip target/${{ matrix.target }}/release/cargo-careful
+
+    - name: Sanity check version
+      run: |
+        VERSION=v$(cargo pkgid | cut -d# -f2)
+        [[ "$VERSION" == "${{ github.ref_name }}" ]] || \
+        echo "Cargo version $VERSION and tag ${{ github.ref_name}} differ" && exit 1
```
Comment (RalfJung) [inline .github/workflows/release.yml:79]: This also doesn't work, needs parentheses around the `echo && exit` it seems.

In the future please either test what you submit as a PR, or state clearly that it has not been tested. Otherwise this creates a bunch of emergency work for maintainers.

**Comment by Luni-4** [comment]
Sorry @RalfJung I haven't had time to test the last changes, I thought they were pretty straightforward looking at other examples around. Since I hadn't provided any example to demonstrate the functionality of this script, I didn't think it would have been immediately used for a seriuous release.
